op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -449,6 +497,140 @@ func runStateMigration(newDBPath string, opts stateMigrationOptions) error {
 	return nil
 }
 
+func runDBCheck(opts dbCheckOptions) (err error) {
+	defer timer("db continuity check")()
+
+	log.Info("DB Continuity Check Started", "dbPath", opts.dbPath)
+
+	ancientDB, err := NewChainFreezer(filepath.Join(opts.dbPath, "ancient"), "", true)
+	if err != nil {
+		return fmt.Errorf("failed to open ancient db: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, ancientDB.Close())
+	}()
+	nonAncientDB, err := openDBWithoutFreezer(opts.dbPath, true)
+	if err != nil {
+		return fmt.Errorf("failed to open non-ancient db: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, nonAncientDB.Close())
+	}()
+
+	lastAncient, err := loadLastAncient(ancientDB)
+	if err != nil {
+		return fmt.Errorf("failed to load last ancient block: %w", err)
+	}
+	lastAncientNumber := lastAncient.Header().Number.Uint64()
+	lastBlockNumber := *rawdb.ReadHeaderNumber(nonAncientDB, rawdb.ReadHeadHeaderHash(nonAncientDB))
+
+	lastBlockNumber = uint64(10000000) // TODO(Alec) remove this line
+
+	var errResult *multierror.Error
+
+	// First, check continuity between ancients and non-ancients.
+	// Gaps in data will often halt the freezing process, so attempting to load the first non-ancient block
+	// will most likely fail if there is a gap.
+	firstNonAncientRange, err := loadNonAncientRange(nonAncientDB, lastAncientNumber+1, 1)
+	if err != nil {
+		if opts.failFast {
+			return fmt.Errorf("failed to load first non-ancient block: %w", err)
+		}
+		// We don't need to add the error to errResult here because it will be added below when we call checkContinuity on non-ancients
+	} else {
+		if _, err := firstNonAncientRange.CheckContinuity(lastAncient, 1); err != nil {
+			err = fmt.Errorf("failed continuity check between ancients and non-ancients: %w", err)
+			if opts.failFast {
+				return err
+			}
+			errResult = multierror.Append(errResult, err)
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(8)
+
+	var mu sync.Mutex
+
+	checkRange := func(start, count uint64, loadRangeFunc func(uint64, uint64) (*RLPBlockRange, error)) {
+		// If we are not at genesis or the first non-ancient block, include the last block of
+		// the previous range so we can check for continuity between ranges.
+		if start != 0 && start != lastAncientNumber+1 {
+			start = start - 1
+			count = count + 1
+		}
+		g.Go(func() error {
+			select {
```

Comment:
@piersy before you dig into the rest, would love a quick check on whether I've done everything right here re using cancel(), ctx etc.